Add Inject Retry Errors On Commit variable

lyang24 opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/86370

Release note (sql change): The session variable
inject_retry_errors_on_commit_enabled has been added. When this is true, any
COMMIT statement will return a transaction retry
error if it is run inside of an explicit transaction. The errors will
continue until inject_retry_errors_on_commit_enabled
is set to false. The purpose of this setting is to allow developers to
test their transaction retry logic.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97226)
<!-- Reviewable:end -->

Commit:
sql: add inject_retry_errors_on_commit_enabled session var

Release note (sql change): The session variable
inject_retry_errors_on_commit_enabled has been added. When this is true, any
COMMIT statement will return a transaction retry
error if it is run inside of an explicit transaction. The errors will
continue until inject_retry_errors_on_commit_enabled
is set to false. The purpose of this setting is to allow developers to
test their transaction retry logic.

lyang24 commented:
I believe tests are flakey how do i rerun failed test?

lyang24 commented:
Hi @rafiss I mimicked your previous pr can you please some feedbacks when you have time thanks!

rafiss reviewed (changes requested):
thanks for your contribution! this is the right direction, and i've left a few comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-:-NOrf5EC3XwoC3VeOSCR:b9jwgn7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @rytaft)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOreVyP01-WNAx7usiZ):*
can you add a release note to the commit message. it should say:

```
Release note (sql change): The session variable
inject_retry_errors_on_commit_enabled has been added. When this is true, any
COMMIT statement will return a transaction retry
error if it is run inside of an explicit transaction. If the client
retries the transaction using the special cockroach_restart SAVEPOINT,
then after the 3rd error, the transaction will proceed as normal.
Otherwise, the errors will continue until inject_retry_errors_enabled
is set to false. The purpose of this setting is to allow developers to
test their transaction retry logic.
```
___
*[`pkg/sql/conn_executor_exec.go` line 939 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOr_aaf18RGYUMkdNXH:-NOr_aaf18RGYUMkdNXI:bb2g0e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_exec.go#L939)):*
> ```Go
> 		GetIdleLatency(ex.statsCollector.PreviousPhaseTimes())
> 
> 	isSetOrShow := ast.StatementTag() == "SET" || ast.StatementTag() == "SHOW"
> ```

i think this should be changed so that the error is only injected if `ast.StatementTag() == "COMMIT"`
___
*[`pkg/sql/conn_executor_exec.go` line 945 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOrd2N10YC9rqx158Le:-NOrd2N10YC9rqx158Lf:b-5hrtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_exec.go#L945)):*
> ```Go
> 				ctx, "injected by `inject_retry_errors_on_commit_enabled` session variable")
> 			return ex.makeErrEvent(retryErr, ast)
> 		}
> ```

similar to the other place where we inject an error, i think this should be changed to be like:

```
	if ex.state.mu.txn.Epoch() < ex.state.lastEpoch+numTxnRetryErrors {
		retryErr := ex.state.mu.txn.GenerateForcedRetryableError(
			ctx, "injected by `inject_retry_errors_on_commit_enabled` session variable")
		return ex.makeErrEvent(retryErr, ast)
	} else {
		ex.state.lastEpoch = ex.state.mu.txn.Epoch()
	}
```

this allows the automatic SAVEPOINT retry logic to work (see my suggestion for a test below)
___
*[`pkg/sql/conn_executor_exec.go` line 1278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOreyrL9oL8NafLGRHn:-NOreyrL9oL8NafLGRHo:b-p47qyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_exec.go#L1278)):*
> ```Go
> 		// numTxnRetryErrors is the number of times an error will be injected if
> 		// the transaction is retried using SAVEPOINTs.
> 		const numTxnRetryErrors = 3
> ```

along with my above comment, can you move the initialization of `numTxnRetryErrors` so that is can be shared between the two functions?
___
*[`pkg/sql/conn_executor_test.go` line 1531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOrardh6vcHMWtW0NC8:-NOrardh6vcHMWtW0NC9:bux2d2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_test.go#L1531)):*
> ```Go
> 
> 	_, err := db.Exec("SET inject_retry_errors_on_commit_enabled = 'true'")
> 	require.NoError(t, err)
> ```

could you add another test cases that uses savepoints, similar to the `with_savepoints` subtest in TestInjectRetryErrors?

and could you add another where both inject_retry_errors_on_commit_enabled and inject_retry_errors_enabled are on?
___
*[`pkg/sql/conn_executor_test.go` line 1553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOrdr0HBoAsa9n5Jm0f:-NOrdr0HBoAsa9n5Jm0g:b-9s6h98) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_test.go#L1553)):*
> ```Go
> 			err = tx.Commit()
> 			if attemptCount >= 5 {
> 				require.NoError(t, err)
> ```

nit: add a `break` here. there's no need to keep running the loop after this
___
*[`pkg/sql/conn_executor_test.go` line 1560 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOreD4v35GGL8HhKXnm:-NOreD4wAFcARwr9iyJj:bfvhrov) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_test.go#L1560)):*
> ```Go
> 				// We should not expect a rollback on commit errors.
> 			}
> 		}
> ```

add an assertion here for `require.Equal(t, 5, txRes)`
___
*[`pkg/sql/conn_executor_test.go` line 1563 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97226#-NOreJPg792BgtPp34OI:-NOreJPg792BgtPp34OJ:b-yblguh) ([raw file](https://github.com/cockroachdb/cockroach/blob/563a54d44e4f57c133f20509131107ce31880d9c/pkg/sql/conn_executor_test.go#L1563)):*
> ```Go
> 	})
> 
> 	t.Run("insert_outside_of_txn", func(t *testing.T) {
> ```

you can remove this insert_outside_of_txn test. it's not needed for explicit transactions


<!-- Sent from Reviewable.io -->

lyang24 commented:
Hi @rafiss thanks for the feedbacks, I have one question about the feedbacks. I am not sure how to do SAVEPOINT test because `crdb.ExecuteTx` uses `ExecuteInTx` method and it will rollback to savepoint on failure. As far as i understood we cannot rollback to previous save point on commit failures because we are already there. example attached below is there any other helper api i can use to testing out the commit retry save point logic?

```
	t.Run("test_injection_failure_on_commit_with_savepoints", func(t *testing.T) {
		// The crdb.ExecuteTx wrapper uses SAVEPOINTs to retry the transaction,
		// so it will automatically succeed.
		var txRes, attemptCount int
		err = crdb.ExecuteTx(ctx, db, nil, func(tx *gosql.Tx) error {
			attemptCount++
			txErr := tx.QueryRow("SELECT $1::int8", attemptCount).Scan(&txRes)
			require.NoError(t, txErr)
			if commitErr := tx.Commit(); commitErr != nil {
				return commitErr
			}
			return nil
		})
		require.NoError(t, err)
		// numTxnRetryErrors is set to 3 in conn_executor_exec, so the 4th attempt
		// should be the one that succeeds.
		require.Equal(t, 4, txRes)
	})
```
	
Failed
=== RUN   TestInjectRetryOnCommitErrors/test_injection_failure_on_commit_with_savepoints
    conn_executor_test.go:1546: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2132/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/conn_executor_test.go:1546
        	Error:      	Received unexpected error:
        	            	restarting txn failed. ROLLBACK TO SAVEPOINT encountered error: sql: transaction has already been committed or rolled back. Original error: pq: restart transaction: TransactionRetryWithProtoRefreshError: injected by `inject_retry_errors_on_commit_enabled` session variable.
        	Test:       	TestInjectRetryOnCommitErrors/test_injection_failure_on_commit_with_savepoints
    --- FAIL: TestInjectRetryOnCommitErrors/test_injection_failure_on_commit_with_savepoints (0.01s)

lyang24 commented:
Hi @rafiss I had some trouble on how to implement retry with save points on commit injected errors. I looked at the code logic even though the errors returned on commit is retryable, the [getRewindTxnCapability](https://github.com/cockroachdb/cockroach/blob/e67e3961a8f4314dd7d92a0bcbe0986f3ce8cee9/pkg/sql/conn_executor.go#L2768) will always return [canAutoRetry = false](https://github.com/cockroachdb/cockroach/blob/e67e3961a8f4314dd7d92a0bcbe0986f3ce8cee9/pkg/sql/conn_executor.go#L2704) and according to the state machine [the transaction will be moved to no transaction state](https://github.com/cockroachdb/cockroach/blob/e67e3961a8f4314dd7d92a0bcbe0986f3ce8cee9/pkg/sql/conn_fsm.go#L287), rollback to savepoint seems to be impossible in this case.
crdb demo example:
```
demo@127.0.0.1:26257/movr  OPEN> commit;                                                                                                                                                                                                                    
ERROR: restart transaction: TransactionRetryWithProtoRefreshError: injected by `inject_retry_errors_on_commit_enabled` session variable
SQLSTATE: 40001
HINT: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html
demo@127.0.0.1:26257/movr>
```

rafiss commented:
Thanks for your explanation! I see the issue now - the auto-retry logic for SAVEPOINTs would only work if the error is injected during the RELEASE TO SAVEPOINT command. However, injecting the error at that point is a bit more complicated.

For now, you can remove the `test_injection_failure_on_commit_with_savepoints` test you added, and the rest of your change should be good.

lyang24 commented:
Hi @rafiss, I have remove the tests and all ci checks looks good do you mind to take another look?

lyang24 commented:
Hi @rafiss I have fixed all the comments do you mind to take another look?

rafiss reviewed (approved):
thanks for your contribution! i will merge this now

bors r+
